django/boss/throttling.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2019 The Johns Hopkins University Applied Physics Laboratory
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from rest_framework.exceptions import Throttled
+
+from django.utils.translation import ugettext_lazy as _
+from django.conf import settings as django_settings
+
+import json
+import bossutils
+import redis
+
+def parse_limit(val):
+    if val is None:
+        return None
+
+    num, unit = val[:-1], val[-1]
+    val = float(num) * {
+        'K': 1024,
+        'M': 1024 * 1024,
+        'G': 1024 * 1024 * 1024,
+        'T': 1024 * 1024 * 1024 * 1024,
+        'P': 1024 * 1024 * 1024 * 1024 * 1024,
+    }[unit.upper()]
+
+    return int(val) # Returning an int, as redis works with ints
+
+class RedisMetrics(object):
+    # NOTE: If there is no throttling redis instance the other methods don't do anything
+    # NOTE: External process will reset values to zero when the window expires
+    # {obj}_metric = current_cost_in_window
+
+    def __init__(self):
+        boss_config = bossutils.configuration.BossConfig()
+        if len(boss_config['aws']['cache-throttle']) > 0:
+            self.conn = redis.StrictRedis(boss_config['aws']['cache-throttle'],
+                                          6379,
+                                          boss_config['aws']['cache-throttle-db'])
+        else:
+            self.conn = None
+
+    def get_metric(self, obj):
+        if self.conn is None:
+            return 0
+
+        key = "{}_metric".format(obj)
+        resp = self.conn.get(key)
+        if resp is None:
+            resp = 0
+        else:
+            resp = int(resp.decode('utf8'))
+        return resp
+
+    def add_metric_cost(self, obj, val):
+        if self.conn is None:
+            return
+
+        key = "{}_metric".format(obj)
+        self.conn.incrby(key, int(val))
+
+class MetricLimits(object):
+    def __init__(self):
+        vault = bossutils.vault.Vault()
+        data = vault.read('secret/endpoint/throttle', 'config')
```

Comment:
How often will Vault be queried? Is it on every request?

Reply:
Yes, because a new `BossThrottle` object is created for each request the limits are read from Vault on each check. This may be more load than wanted.

@movestill / @sandyhider Should Vault data be cached for a minimal period of time?